tests/lexer/tstrnumlits.nim

```diff
@@ -0,0 +1,151 @@
+# Test tkStrNumLit
+
+import std/macros
+import strutils
+import mlexerutils
+
+type
+  S = distinct string
+
+proc `'wrap`(number: string): S =
+  result = ("[[" & number & "]]").S
+
+assertAST dedent """
+  StmtList
+    ProcDef
+      AccQuoted
+        Ident "\'wrap"
+      Empty
+      Empty
+      FormalParams
+        Ident "S"
+        IdentDefs
+          Ident "number"
+          Ident "string"
+          Empty
+      Empty
+      Empty
+      StmtList
+        Asgn
+          Ident "result"
+          DotExpr
+            Par
+              Infix
+                Ident "&"
+                Infix
+                  Ident "&"
+                  StrLit "[["
+                  Ident "number"
+                StrLit "]]"
+            Ident "S"""":
+  proc `'wrap`(number: string): S =
+    result = ("[[" & number & "]]").S
+
+
+template `'twrap`(number: string): untyped =
+  number.`'wrap`
+
+proc extraContext(): S =
+  22.40'wrap
+
+# if a user is writing a numeric library, they might write op procs similar to:
+
+proc `*`(left, right: S): S =
+  result = (left.string & "times" & right.string).S
+
+proc `+`(left, right: S): S =
+  result = (left.string & "plus" & right.string).S
+
+proc `$`(number: S): string =
+  result = number.string
+
+doAssert string(1'wrap) == "[[1]]":
+  "unable to resolve an integer-suffix pattern"
+
+doAssert string(-1'wrap) == "[[-1]]":
+  "unable to resolve a negative integer-suffix pattern"
+
+doAssert lispReprStr(-1'wrap) == """(DotExpr (StrLit "-1") (Ident "\'wrap"))""":
+  "failing to build correct AST on `-1'wrap`."
+
+doAssert string(12345.67890'wrap) == "[[12345.67890]]":
+  "unable to resolve a float-suffix pattern"
+
+doAssert string(1'wrap*1'wrap) == "[[1]]times[[1]]":
+  "unable to resolve an operator between two suffixed numeric literals"
+
+doAssert string(1'wrap+ -1'wrap) == "[[1]]plus[[-1]]":
+  "unable to resolve a negative suffixed numeric literal following an operator"
+
+doAssert string(1'twrap) == "[[1]]":
+  "unable to resolve a template using a suffixed numeric literal"
+
+doAssert string(extraContext()) == "[[22.40]]":
+  "unable to return a suffixed numeric literal by an implicit return"
+
+doAssert string(0x5a3a'wrap) == "[[0x5a3a]]":
```

Comment:
also add an example showing case is preserved eg:
```nim
0xdeadBEEF
0.1E12
0.0e12
0.0e+12
0.0e-12
```

Reply:
Done.